### DISCUSSION TITLE
Modify matching behavior when it fails to find the matching character

### DISCUSSION THREAD

**Discussion by tapyu** [helix-editor/helix#10567] [Ideas]
I frequently find myself in the following situation:

```
|esc = ["collapse_selection", "keep_primary_selection"]
```

where `|` denotes the cursor position. I want to match in next inner brackets content, that is, `"collapse_selection", "keep_primary_selection"`. I am not a pro`helix` user, but AFAIK you need to run `f[` and `mi[` in order to select it, which is not that effective.

I propose to extend the match mode functionalities to cover this situation:
1. The bind should be `mf` + any character + one of the [match mode binding](https://docs.helix-editor.com/keymap.html#match-mode).
1. The command should apply `find_next` command to this next character.
2. Run `match_brackets`,  `surround_add`,`surround_delete`, `select_textobject_around`,  `select_textobject_inner`, `surround_replace` depending whether `m`, `s`,  `d`,  `a`, `i`, or `r` has been selected.
3. Another bind `mF` must be considered to search for a matching before the current cursor.

I tried to create a keybind in the config file for this case, but I failed. Moreover, I do think that this situation is recurrent, so other users might also get this functionality out of the box.

Finally, in my previous example, `mi[` does nothing, which is pretty stupid. I propose that, once the matching fails to find a pair, it must fallback to my proposed keybind `mf[i` (or `mF[i`) in order to find the next (or previous) matching.

If you find `mf[i` or `mF[i` too complicated for a single command, I at least propose to modify the behavior of the matching mode when it fails to find matching inner/around/whatever character. To do anything is better than to do nothing.

**Comment by tapyu** [comment]
Just to let you know, my trial to create a binding for this was

```
m = { f = { i = ["find_next_char", "select_textobject_inner"] } }
```

**Comment by thomasschafer** [comment]
I would also be in favour of adding something like this, similar to the `around_next` and `inside_next` commands in the nvim plugin [mini.ai](https://github.com/echasnovski/mini.ai), which in the example you shared above would select the contents of the array with `vin[`.

I'm going to have a crack at implementing this myself.

**Comment by thomasschafer** [comment]
> 3. Run `match_brackets`,  `surround_add`,`surround_delete`, `select_textobject_around`,  `select_textobject_inner`, depending whether `m`, `s`,  `d`,  `a`, or `i` has been selected.

I'm confused by this - how would `surround_add` and `surround_delete` work? I don't see how it would make sense to run a command like `mfi(` - where would that add the `(`?

**Reply by tapyu** [reply]
I will use
```
|esc = ["collapse_selection", "keep_primary_selection"]
```
as example for all your questions (`|` represents the cursor position).

>  how would `surround_delete` work?

`mfd[` would produce
```
esc = |"collapse_selection", "keep_primary_selection"
```
and `mfd]` would produce
```
esc = "collapse_selection", "keep_primary_selection"|
```
In words, `mfd[` deletes the next matching brackets pair.

> I don't see how it would make sense to run a command like `mfi(` - where would that add the `(`?

`mfi(` doesn't add anything, it **selects** inside an object. In my example, `mfi[` would select what it is in bold

esc = [**"collapse_selection", "keep_primary_selection"**]


>  how would `surround_add` work?

Okay, I give you that, `surround_add` is the unique option that makes no sense to be implemented with `mf` or `mF` because is adds things with respect to a previous selection. Therefore, my reviewed proposal would be only the bindings `mfm` (`match_brackets`), `mfr` (`surround_replace`), `mfd` (`surround_delete`), `mfa` (`select_textobject_around`), `mfi` (`select_textobject_inner`). The same goes for `mF`.

**Reply by thomasschafer** [reply]
Apologies, `mfi(` was a typo, I meant to write `mfs(`, but you answered that anyway! That all makes sense to me.

My only other question would be whether `f` and `F` would be the right keys to use, or whether something like `n` and `p` (meaning "next" and "previous") would be better, with my slightly tenous arguments being

- [mini.ai](https://github.com/echasnovski/mini.ai) (which is built in to the popular [LazyVim](https://github.com/LazyVim/LazyVim)) uses `n` already, and users might already be used to that
- typing `F` requires two keys presses (`shift` + `f`) compared to just one for `p`
- semantically I think "match next inner \<object\>" makes more sense than "match find inner \<object\>"

There is one major downside to using `n` instead of `f`, however, which is that typing `mn` quickly is hard because most people would use their right index for both keys. However, this would no longer be a problem if it was decided that the ordering should be `min<object>` rather than `mni<object>` (and similarly for other commands), and from a quick dig around I think the implementation would be more straightforward with the former option.

**Reply by tapyu** [reply]
> My only other question would be whether f and F would be the right keys to use, or whether something like n and p (meaning "next" and "previous") would be better, with my slightly tenous arguments being

That is open to discussion, indeed. My idea was to leverage the `f` and `F` behavior to go afterwards and backwards to create the `mf` and `mF` keybindings with similar behavior.

> typing F requires two keys presses (shift + f) compared to just one for p

Agreed.

> There is one major downside to using n instead of f, however, which is that typing mn quickly is hard because most people would use their right index for both keys.

I have no idea what "right index" is, so I couldn't understand what the problem with `mni<object>` is. It seems great to me.

**Reply by thomasschafer** [reply]
> I have no idea what "right index" is, so I couldn't understand what the problem with mni\<object\> is. It seems great to me.

Ah, I mean right index finger - typing with the "correct" fingers on a qwerty keyboard leads to the right index finger being used for both m and n, although I appreciate that some people use non-qwerty layouts, or just use different fingers

**Reply by tapyu** [reply]
Ah, I am sorry for my stupidity. I got it now. However, I wouldn't assume that everyone who is on a qwerty keyboard will use the right index twice to type `m` and `n`. I see myself using the middle and index finger for that, and I also use a qwerty keyboard.

**Reply by tapyu** [reply]
A very important aspect about helix. In this situation
```
|esc = ["collapse_selection", "keep_primary_selection"]
```
pressing `mi[` does nothing. This is stupid. A waste of keys. When a user runs a command, they expect something: **doing something is better than doing nothing**. Therefore, I propose `helix` to fallback to `mfi[` if `mi[` matches nothing.

It is not only effective but also reasonable, if the user is trying to match something and `hx` couldn't resolve this matching at the current position, the most reasonable solution is to try to find the next matching. Doing nothing is not a reasonable solution, at all.

**Comment by thomasschafer** [comment]
Created an [initial PR](https://github.com/helix-editor/helix/pull/11124) implementing finding in next character pair - I'll add to it if we can get an agreement on the approach and implementation

**Comment by thomasschafer** [comment]
Linking [this](https://github.com/helix-editor/helix/discussions/9895) which discusses essentially the same idea but with different keys used
